File: dialogio/Menu1_refresh.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

static void Menu1_refresh (void)
{
  DBG_ENTER ("Menu1_refresh");

  if ((_mnu1.page != _mnu1.opage) || (_mnu1.idx2 != _mnu1.opos) || mn1_environment.do_refresh)
  {
    _mnu1.opage = _mnu1.page;
    _mnu1.opos  = _mnu1.idx2;
    mn1_environment.do_refresh = false;

    for (_mnu1.idx = _mnu1.page; _mnu1.idx <= _mnu1.len2 + _mnu1.page - 1; _mnu1.idx++)
    {
      if (_mnu1.idx == _mnu1.idx2 + _mnu1.page - 1)
      {
        String_t s, t;

        s = Menu1_pstr (_mnu1.idx2 + _mnu1.page - 1);
        t = Menu1_pdes (_mnu1.idx2 + _mnu1.page - 1);
        AppendString ((String *)&s, (String *)&t, 255);
        s = ExpStrR ((String *)&s, _mnu1.max + s.len - CStrLen ((String *)&s), ' ');

        Menu1_ShowCStr (mn1_environment.v_dest,
                        _mnu1.x + 1,
                        _mnu1.y + _mnu1.idx2,
                        (String *)&s,
                        mn1_setting.text2_attr,
                        mn1_setting.short2_attr,
                        mn1_setting.text_attr,
                        mn1_setting.short_attr);
      }
      else
      {
        String_t s, t;
        size_t sl;
        uint8_t attrs[2];

        s = Menu1_pstr (_mnu1.idx);
        sl = s.len;
        t = Menu1_pdes (_mnu1.idx);
        AppendString ((String *)&s, (String *)&t, 255);
        t = Menu1_pdes (_mnu1.idx2 + _mnu1.page - 1);
        s = ExpStrR ((String *)&s, _mnu1.max + sl + t.len - CStrLen ((String *)&s), ' ');

        if (_mnu1.mbuf[_mnu1.idx - 1].use)
        {
          attrs[0] = mn1_setting.text_attr;
          attrs[1] = mn1_setting.short_attr;
        }
        else
        {
          attrs[0] = mn1_setting.disbld_attr;
          attrs[1] = mn1_setting.disbld_attr;
        }

        ShowCStr (mn1_environment.v_dest,
                  _mnu1.x + 1,
                  _mnu1.y + _mnu1.idx - _mnu1.page + 1,
                  (String *)&s,
                  attrs[0],
                  attrs[1]);
      }
    }

    _mnu1.vscrollbar_pos =
      VScrollBar (mn1_environment.v_dest,
                  _mnu1.x + _mnu1.max + 1,
                  _mnu1.y + 1,
                  _mnu1.vscrollbar_size,
                  _mnu1.count,
                  _mnu1.idx2 + _mnu1.page - 1,
                  _mnu1.vscrollbar_pos,
                  mn1_setting.menu_attr,
                  mn1_setting.menu_attr);
  }

  DBG_LEAVE (); //EXIT //Menu1_refresh
}

File: dialogio/pas/Menu1_refresh.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// From `MenuLib1' unit

procedure refresh;

{$I Menu1_ShowCStr.pas}

begin
  _dbg_enter ({$I %FILE%}, 'refresh');

  If (page = opage) and (idx2 = opos) and NOT MenuLib1_mn_environment.do_refresh then
    begin
      _dbg_leave; EXIT; //refresh
    end
  else begin
         opage := page;
         opos  := idx2;
         MenuLib1_mn_environment.do_refresh := FALSE;
       end;

  For idx := page to mnu_len2+page-1 do
    If (idx = idx2+page-1) then
      ShowCStr_clone(MenuLib1_mn_environment.v_dest,mnu_x+1,mnu_y+idx2,
                     ExpStrR(pstr(idx2+page-1)+pdes(idx2+page-1),
                     max+(Length(pstr(idx2+page-1))+Length(pdes(idx2+page-1))-
                     CStrLen(pstr(idx2+page-1)+pdes(idx2+page-1))),' '),
                     MenuLib1_mn_setting.text2_attr,MenuLib1_mn_setting.short2_attr,
                     MenuLib1_mn_setting.text_attr,MenuLib1_mn_setting.short_attr)
    else
      If mbuf[idx].use then
        ShowCStr(MenuLib1_mn_environment.v_dest,mnu_x+1,mnu_y+idx-page+1,
                 ExpStrR(pstr(idx)+pdes(idx),
                 max+(Length(pstr(idx))+Length(pdes(idx2+page-1))-
                 CStrLen(pstr(idx)+pdes(idx))),' '),
                 MenuLib1_mn_setting.text_attr,MenuLib1_mn_setting.short_attr)
      else
        ShowCStr(MenuLib1_mn_environment.v_dest,mnu_x+1,mnu_y+idx-page+1,
                 ExpStrR(pstr(idx)+pdes(idx),
                 max+(Length(pstr(idx))+Length(pdes(idx2+page-1))-
                 CStrLen(pstr(idx)+pdes(idx))),' '),
                 MenuLib1_mn_setting.disbld_attr,MenuLib1_mn_setting.disbld_attr);

  vscrollbar_pos :=
    VScrollBar(MenuLib1_mn_environment.v_dest,mnu_x+max+1,mnu_y+1,temp2,mnu_count,idx2+page-1,
               vscrollbar_pos,MenuLib1_mn_setting.menu_attr,MenuLib1_mn_setting.menu_attr);

  _dbg_leave; //EXIT //refresh
end;
